hc_dpki/src/keypair.rs

```diff
@@ -0,0 +1,562 @@
+        })
+    }
+
+    /// initialize the pairs from an encrypted persistence bundle
+    ///
+    /// @param {object} bundle - persistence info
+    ///
+    /// @param {SecBuf} passphrase - decryption passphrase
+    pub fn from_bundle(
+        bundle: &bundle::KeyBundle,
+        passphrase: &mut SecBuf,
+    ) -> Result<Keypair, HolochainError> {
+        // decoding the bundle.data of type util::ReturnBundledata
+        let bundle_decoded = base64::decode(&bundle.data)?;
+        let bundle_string = str::from_utf8(&bundle_decoded).unwrap();
+        let data: bundle::ReturnBundleData = json::decode(&bundle_string).unwrap();
+        let mut decrypted_data = SecBuf::with_secure(BUNDLE_DATA_LEN);
+        util::pw_dec(&data, passphrase, &mut decrypted_data)?;
+        let mut sign_priv = SecBuf::with_secure(64);
+        let mut enc_priv = SecBuf::with_secure(32);
+
+        let pub_keys = {
+            let decrypted_data = decrypted_data.read_lock();
+            if decrypted_data[0] != 1 {
+                return Err(HolochainError::ErrorGeneric(format!(
+                    "Invalid Bundle Version : {:?}",
+                    decrypted_data[0]
+                )));
+            }
+            sign_priv.write(0, &decrypted_data[65..129])?;
+            enc_priv.write(0, &decrypted_data[129..161])?;
+
+            KeyBuffer::with_raw_parts(
+                array_ref![&decrypted_data, 1, 32],
+                array_ref![&decrypted_data, 33, 32],
+            )
+            .render()
+        };
+
+        Ok(Keypair {
+            pub_keys,
+            enc_priv,
+            sign_priv,
+        })
+    }
+
+    /// sign some arbitrary data with the signing private key
+    ///
+    /// @param {SecBuf} data - the data to sign
+    ///
+    /// @param {SecBuf} signature - Empty Buf the sign
+    pub fn sign(
+        &mut self,
+        data: &mut SecBuf,
+        signature: &mut SecBuf,
+    ) -> Result<(), HolochainError> {
+        sign::sign(data, &mut self.sign_priv, signature)?;
+        Ok(())
+    }
+
+    /// verify data that was signed with our private signing key
+    ///
+    /// @param {SecBuf} signature
+    ///
+    /// @param {SecBuf} data
+    pub fn verify(
+        &mut self,
+        signature: &mut SecBuf,
+        data: &mut SecBuf,
+    ) -> Result<i32, HolochainError> {
+        let mut sign_pub = SecBuf::with_insecure(sign::PUBLICKEYBYTES);
+        let mut enc_pub = SecBuf::with_insecure(kx::PUBLICKEYBYTES);
+
+        util::decode_id(self.pub_keys.clone(), &mut sign_pub, &mut enc_pub)?;
+        let v: i32 = sign::verify(signature, data, &mut sign_pub);
+        Ok(v)
+    }
+
+    /// encrypt arbitrary data to be readale by potentially multiple recipients
+    ///
+    /// @param {array<string>} recipientIds - multiple recipient identifier strings
+    ///
+    /// @param {Buffer} data - the data to encrypt
+    ///
+    /// @param {Buffer} out - Empty vec[secBuf]
+    pub fn encrypt(
```

Comment:
Same comment as with `verify`: how can we encrypt something only holding the public key?

Reply:
This encrypt function is something of a POC. It will likely be implemented slightly differently, and as you say, may be moved into AgentId